### ISSUE TITLE
Improve branch handling after upgrade

### ISSUE THREAD

**Issue by ripleymj** [jmunixusers/cs-vm-build#211]
Handling Mint upgrades is a huge conversation for us, but I'd like to propose one piece.

In our wrapper, could a check be added to warn (or offer to replace) the currently chosen branch when:
- url is set to github.com/jmunixusers/cs-vm-build
- branch != os-release

The branch would then be set to the value from os-release, perhaps after a git probe to see if it exists, and master if that fails. Checking these two conditions should prevent unnecessary warnings for us when developing, and attempt to improve Ansible handling for people who choose to upgrade their VM.

It is not a goal of this issue to attempt to upgrade VMs.

**Comment by kylelaker** [comment]
I would be okay with this. In fact, I'd further support a special warning for master.

Incorrect release:

> You are using a version of the configuration tool meant for a different release of
Linux Mint. Would you like to switch to the proper release?

Master warning:

> You are currently an unstable development branch of the configuration tool.
Would you like to switch to the proper stable release?

**Comment by ripleymj** [comment]
Any value in attempting lexicographic comparison of versions, such as:
```
if(os-release > chosenBranch)
    warn("Please upgrade branches")
else if (chosenBranch > os-release)
   warn("Your life has taken a wrong turn")
```

again, only for the jmunixusers repo.

**Comment by kylelaker** [comment]
I definitely support that

**Comment by kylelaker** [comment]
Actually, I am unsure if we are guaranteed that codenames will be lexicographically greater than those of previous releases. Mint 17.1 was Rebecca and Mint 17.2 was Rafaela (and `'rebecca' > 'rafaela'`). This is the only occurrence of this in the history of Mint that I can find, but as they approach the end of the alphabet, who knows what will happen.

**Comment by kylelaker** [comment]
Targeting to Spring 2019 in the hopes of us getting this right for Tessa.

**Comment by ripleymj** [comment]
@cfrome77, @kylelaker is attempting some major changes to the Python script, so you may want to double check with him before investing too much time in this.

**Comment by kylelaker** [comment]
I do plan to make some significant changes to the script; however, I do not want that work to prevent us from making meaningful user-facing changes. My eventual goal is to separate the tool into a few modules and clean up the code. That can be done no matter what gets done to the existing script. I will make sure that I re-incorporate any changes we see before I finish my work (*if* I finish my work).

**Comment by cfrome77** [comment]
@kylelaker I have a working version that warns the user about using an outdated branch or about using master, and suggests the branch to use. It does not hinder the code from running, only warns the user before running the code. I will summit a PR for you to take a look.

**Comment by kylelaker** [comment]
@cfrome77 That sounds good. I have a few other ideas for this that we might be able to incorporate into this PR or into another one depending on what your thoughts are.

Basically, there are actually a few things I'd like to detect for this. We can guess that a release is Mint release if it matches `[a-z]+a` since all Linux Mint release are female names ending in `a` and they only contain `[a-z]`. Most other branches we see do match this. We also already shell out to `git ls-remote` and I think we can continue doing this. Therefore I think we can do something like the following:

```python
system_version = get_distro_release_name()
chosen_branch = USER_CONFIG['git_branch']
branch_mismatch = system_version != chosen_branch
looks_minty = re.compile(r"[a-z]+a").fullmatch(release)
ls_remote = subprocess.run(
    ["/usr/bin/git", "ls-remote", '--heads', USER_CONFIG['git_url']],
    stdout=subprocess.PIPE
)
remote_refs = []
for ref in ls_remote:
    # Get the full ref name (such as refs/heads/master)
    hash_ref_list = ref.split('\t')[0]
    # Get the branch name (such as master)
    remote_refs.append(hash_ref_list[1].split("/")[-1])
system_exists = system_version in remote_refs
chosen_exists = chosen_branch in remote_refs
```

The conditions I'd like to check for are:

 - `mismatch and looks_minty and system_exists`: Suggest that the user switch to the correct branch in settings since they likely upgraded Mint but have a config saved with an 'old' branch
 - `mismatch and looks_mint and not system_exists`: The user is either on a really old Mint release or they upgraded to a release we don't yet have a branch for. We should suggest master.
 - `system_exists and not chosen_exists`: The user has selected a branch that does not exist on the remote they chose---they need to either fix their remote or switch to a different branch
 - `(not system_exists) and (not chosen_exists)`: We do not support the user's OS and the branch chosen does not exist on their chosen remote. The user should probably use master.
 - system_exists and `branch == 'master'`: Maybe we should give a warning here? The user probably should go ahead and switch to the branch for their version of mint.

We could also add a condition for some of the above to see whether the user has selected jmunixusers/cs-vm-build or if they are using their own fork. Perhaps a new global called `DEFAULT_GIT_REMOTE` and we can check `DEFAULT_GIT_REMOTE != USER_CONFIG['git_url']`.

Let me know your thoughts on these checks -- whether they're too much, not enough, or something else.

**Comment by kylelaker** [comment]
@cfrome77 I see you've submitted a PR. I will review that as well.

**Comment by cfrome77** [comment]
@kylelaker  I definitely like having all these checks, and think they do a good job of catching every mistake a user could make related to branch and remote in the config. I know my PR covers some but not all of these. I will see what I can do about adding the other checks.

**Comment by kylelaker** [comment]
Thanks, Chris! I really appreciate your flexibility with this and for taking on this issue. Let me know if it starts to look like it might be a bit too much for a single PR or if you have any questions.

**Comment by cfrome77** [comment]
@kylelaker What was your thinking for DEFAULT_GIT_REMOTE != USER_CONFIG['git_url']. I am confused on which cases this would be used. Do we only want to warn the use if the url specified is the jmunixusers repsoitory or should the warnings be shown for url's of forked copies as well.

**Comment by kylelaker** [comment]
@cfrome77 That's a great question. I think we can assume that if the user has been brave enough to use a non-default repo, then they probably don't need our helpful warnings. It would also prevent a warning if, for example, I were a testing a fix on a branch called `fixthinga` (which isn't a great branch name but possible).

So yeah, I'd like to see most of the occurrences of `'https://github.com/jmunixusers/cs-vm-build'` changed to a variable (something like `DEFAULT_GIT_REMOTE`) and I'd like to ensure that user's chosen URL is not the same as the default URL variable (`USER_CONFIG['git_url'] != DEFAULT_GIT_REMOTE`) or the inverse condition if that makes the code cleaner.

So basically, we don't want to show the warning for forks. Let me know if I didn't answer your question sufficiently.

**Comment by cfrome77** [comment]
@kylelaker I have a working version. The only thing I see is that with the check for `looks_minty` any branch name ending in `a`, such as your example of `fixthinga` will generate the warning since anything ending in a satisfies the condition of the regex even if it is not actually a name of a mint release. I will commit my changes since that is the only thing I see at the moment.

**Comment by kylelaker** [comment]
This was fixed by #223
